src/scheduler.cpp

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "scheduler.h"
+
+#include <assert.h>
+#include <boost/bind.hpp>
+#include <utility>
+
+CScheduler::CScheduler() : nThreadsServicingQueue(0)
+{
+}
+
+CScheduler::~CScheduler()
+{
+    assert(nThreadsServicingQueue == 0);
+}
+
+
+#if BOOST_VERSION < 105000
+static boost::system_time toPosixTime(const boost::chrono::system_clock::time_point& t)
+{
+    boost::chrono::system_clock::duration d = t.time_since_epoch();
+    boost::chrono::microseconds usecs = boost::chrono::duration_cast<boost::chrono::microseconds>(d);
+    boost::system_time result = boost::posix_time::from_time_t(0) +
+        boost::posix_time::microseconds(usecs.count());
+    return result;
+}
+#endif
+
+void CScheduler::ServiceQueue()
+{
+    boost::unique_lock<boost::mutex> lock(newTaskMutex);
+    ++nThreadsServicingQueue;
+
+    // newTaskMutex is locked throughout this loop EXCEPT
+    // when the thread is waiting or when the user's function
+    // is called.
+    while (1) {
+        try {
+            while (taskQueue.empty()) {
+                // Wait until there is something to do.
+                newTaskScheduled.wait(lock);
+            }
+            // Wait until either there is a new task, or until
+            // the time of the first item on the queue:
+
+            // wait_until needs boost 1.50 or later. For older boosts, convert to
+            // posix time:
+#if BOOST_VERSION < 105000
+            while (newTaskScheduled.timed_wait(lock, toPosixTime(taskQueue.begin()->first)) == true) {
+                // Keep waiting until timeout
+            }
+#else
+            while (newTaskScheduled.wait_until(lock, taskQueue.begin()->first) != boost::cv_status::timeout) {
+                // Keep waiting until timeout
+            }
+#endif
+            // If there are multiple threads, the queue can empty while we're waiting (another
+            // thread may service the task we were waiting on).
+            if (taskQueue.empty()) continue;
```

Comment:
Mmm. I convinced myself an empty check isn't necessary there, but belt&suspenders says I should just put it back.

(not needed because: if the queue is emptied during timeout, then the wait loop is timed out and exited and the empty queue is caught here. And since there is no unschedule-task, the only other way the wait can exit is if a new task is added to the queue, which wakes up only one thread and so the queue can't be empty).

I'll add the empty condition to the wait loops.